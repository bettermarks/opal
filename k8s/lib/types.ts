import { ResourceRequirements } from "../imports/k8s";
/**
 * Stages
 */
export enum Stage {
  LOC = "loc",
  DEV = "dev",
  CI = "ci",
  PRO = "pro",
}

/**
 * Segments
 */
export enum Segment {
  LOC00 = "loc00",
  DEV00 = "dev00",
  CI00 = "ci00",
  CI01 = "ci01",
  PRO00 = "pro00",
}

/**
 * Defines the type defnition of labels in key-value format
 * - Applied to Nodepools when creating K8S cluster
 * - Applied to Deployment definitions at pod spec level when creating a deployment
 */
export type NodeSelector = {
  nodetype: string;
};

export type ApplicationConfig = {
  SEGMENT: string;
  LOG_FORMAT: string;
  LOG_LEVEL: LogLevel;
  LICENSING_SERVICE_URL: string;
  APM_URL: string;
  APM_ENABLED: boolean;
  APM_TRANSACTION_SAMPLE_RATE: string;
  JWT_VERIFICATION_KEYS: { [key: string]: object };
  EVENTS_EXPORT_FUNCTION?: string;
  EVENTS_EXPORT_EXPORT_HOOK?: string;
  DATA_EVENT_API_URL?: string;
  SDWH_PORT?: string;
  SDWH_USER?: string;
  SDWH_DB_HOST?: string;
  SDWH_DB_USER?: string;
  SDWH_DB_PORT?: string;
  SDWH_DB_NAME?: string;
};

/**
 * Deployment configuration
 */
export type DeploymentConfig = {
  /**
   * Migration job resources
   */
  migrationJobResources: ResourceRequirements;
  /**
   * Load fixtures job resources
   */
  loadFixturesJobResources: ResourceRequirements;
  /**
   * API resources
   */
  apiResources: ResourceRequirements;
  /**
   * API replicas
   */
  apiReplicas: number;
  /**
   * event export job resources
   */
  eventExportResources: ResourceRequirements;
};

/**
 * Log level
 */
export enum LogLevel {
  INFO = "INFO",
  WARNING = "WARNING",
  ERROR = "ERROR",
  DEBUG = "DEBUG",
}

/**
 * Namespaces
 */
export enum Namespace {
  DEFAULT = "default",
  LICENSING = "licensing",
}
