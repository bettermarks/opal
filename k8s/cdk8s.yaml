language: typescript
app: ts-node main.ts
imports:
  - k8s@1.27.0
