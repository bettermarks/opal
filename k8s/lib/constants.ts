import { Toleration } from "../imports/k8s";
import { NodeSelector } from "./types";

export const APP_NODE_POOL_LABELS: NodeSelector = {
  nodetype: "application",
};

// Default tolerations
export const DEFAULT_TOLERATIONS: Toleration[] = [
  {
    key: "node.kubernetes.io/unreachable",
    operator: "Exists",
    effect: "NoExecute",
    tolerationSeconds: 20,
  },
  {
    key: "node.kubernetes.io/not-ready",
    operator: "Exists",
    effect: "NoExecute",
    tolerationSeconds: 20,
  },
];

/**
 * Name of the secret which is used by LicensingChart and MigrationJobChart to look up secrets.
 */
export const LICENSING_SECRET = "licensing";
/**
 * Name of the secret which is used by EventExportCronJob chart to look up secrets.
 */
export const EVENT_EXPORT_SECRET = "event-export";

/**
 * Name of the secret used by Service Account to pull image from AWS ECR registry.
 * This secret is created as part of a K8s cluster setup in bm-operations repo.
 */
export const REGISTRY_CREDENTIALS = "registry-credentials";

/**
 * Name of the image used for postgres database.
 */
export const POSTGRES_IMAGE = "postgres:14";

/**
 *  Local postgres
 */
export const LOCAL_POSTGRES_ENV_DATA = {
  DB_HOST: "licensing-db",
  DB_PORT: "5432",
  DB_USER: "postgres",
  DB_PASSWORD: "postgres",
  DB_NAME: "postgres",
  POSTGRES_PASSWORD: "postgres",
  POSTGRES_USER: "postgres",
  POSTGRES_DB: "postgres",
};

/**
 * External Secrets Cluster Secret Store: STACKIT Secrets Manager (Vault)
 */
export const ESO_STACKIT_SECRETS_MANAGER = "stackit-secrets-manager";
