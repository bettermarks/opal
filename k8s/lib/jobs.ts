import { Container, KubeJob } from "../imports/k8s";
import { Construct } from "constructs";
import { NodeSelector } from "./types";
import { RestartPolicy } from "cdk8s-plus-27";

export type LicensingJobProps = {
  name: string;
  namespace: string;
  ttlSecondsAfterFinished: number;
  nodeSelector?: NodeSelector;
  serviceAccountName: string;
  initContainers: Container[];
  containers: Container[];
};

export class LicensingJob extends Construct {
  readonly job: KubeJob;
  constructor(scope: Construct, id: string, props: LicensingJobProps) {
    super(scope, id);
    /**
     * Node selector
     */
    let nodeSelector: { [key: string]: string } = {};
    if (props.nodeSelector) {
      nodeSelector[nodeSelector.nodeLabelKey] = nodeSelector.nodeLabelValue;
    }

    this.job = new KubeJob(this, `${props.name}`, {
      metadata: {
        name: props.name,
        namespace: props.namespace,
      },
      spec: {
        backoffLimit: 1, // number of retries
        ttlSecondsAfterFinished: props.ttlSecondsAfterFinished, // delete job after specified time has finished
        template: {
          spec: {
            serviceAccountName: props.serviceAccountName,
            nodeSelector: nodeSelector,
            restartPolicy: RestartPolicy.NEVER,
            initContainers: props.initContainers,
            containers: props.containers,
          },
        },
      },
    });
  }
}
