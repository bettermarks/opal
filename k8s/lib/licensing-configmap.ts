import { ConfigMap } from "cdk8s-plus-27";
import { Construct } from "constructs";
import { ApplicationConfig } from "./types";
import { Chart, ChartProps } from "cdk8s";

export type LicensingConfigProps = ChartProps & {
  name: string;
  appConfig: ApplicationConfig;
  namespace?: string;
};

export class LicensingConfig extends Chart {
  readonly configMap: ConfigMap;

  constructor(scope: Construct, id: string, props: LicensingConfigProps) {
    super(scope, id);
    const { name, namespace, appConfig } = props;

    this.configMap = new ConfigMap(this, name, {
      metadata: {
        name,
        namespace,
      },
    });
    for (const [key, value] of Object.entries(appConfig)) {
      if (
        Array.isArray(value) ||
        typeof value === "boolean" ||
        typeof value === "object"
      ) {
        this.configMap.addData(key, JSON.stringify(value));
      } else {
        this.configMap.addData(key, value);
      }
    }
  }
}
