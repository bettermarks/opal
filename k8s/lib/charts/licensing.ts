import { Chart, ChartProps } from "cdk8s";
import { HttpIngressPathType, ImagePullPolicy } from "cdk8s-plus-27";
import { Construct } from "constructs";
import {
  EnvFromSource,
  IntOrString,
  KubeIngress,
  ResourceRequirements,
} from "../../imports/k8s";
import { LICENSING_INGRESS_CLASS } from "../constants";

import { LicensingService } from "../services/licensing-service";
import { Namespace, NodeSelector, Segment } from "../types";

/**
 * This class is the implementation detail of Licensing deployment.
 */
export type LicensingChartProps = ChartProps & {
  /**
   * Docker image for Licensing
   */
  image: string;
  /**
   * Name of the secret containing Postgres credentials
   *
   * Required keys:
   * - DB_HOST
   * - DB_PORT
   * - DB_USER
   * - DB_PASSWORD
   * - DB_NAME
   */
  postgresSecret: string;
  /**
   * Name of the secret containing necessary credentials for running the service.
   *
   * Required keys:
   * - APM_SECRET_TOKEN
   * - LICENSING_SERVICE_KID
   * - SHOP_SERVICE_KID (will be outsourced to shop service)
   * - SHOP_INVOICE_AUTH_PASSWORD (will be outsourced to shop service)
   */
  applicationSecret?: string;
  /**
   * Name of the secret containing the licensing service private key.
   *
   * Required key:
   * - LICENSING_SERVICE_PRIVATE_KEY
   */
  licensingServicePrivateKey?: string;
  /**
   * Name of the secret containing the shop service private key.
   *
   * Required key:
   * - SHOP_SERVICE_PRIVATE_KEY
   */
  shopServicePrivateKey?: string;
  serviceAccountName: string;
  segment: Segment;
  name: string;
  /**
   * Name of the config map containing necessary configuration.
   * Keys are derived from the type ApplicationConfig
   */
  configMap: string;
  nodeSelector?: NodeSelector;
  apiResources?: ResourceRequirements;
  /**
   * API replicas
   * @default 3
   */
  apiReplicas?: number;
};

/**
 * Deploys Licensing server.
 */
export class LicensingChart extends Chart {
  constructor(scope: Construct, id: string, props: LicensingChartProps) {
    super(scope, id, props);

    const {
      namespace = Namespace.LICENSING,
      apiReplicas = 1,
      apiResources,
      applicationSecret,
      licensingServicePrivateKey,
      shopServicePrivateKey,
      image,
      nodeSelector,
      postgresSecret,
      segment,
      serviceAccountName,
      name,
      configMap,
    } = props;

    const applicationEnv: EnvFromSource[] = [
      {
        configMapRef: {
          name: configMap,
        },
      },
      {
        secretRef: {
          name: postgresSecret,
        },
      },
      {
        secretRef: {
          name: licensingServicePrivateKey,
        },
      },
      {
        secretRef: {
          name: shopServicePrivateKey,
        },
      },
      ...(applicationSecret
        ? [
            {
              secretRef: {
                name: applicationSecret,
              },
            },
          ]
        : []),
    ];

    const apiName = `${name}-api`;
    const licensingService = new LicensingService(this, `${apiName}-service`, {
      name: apiName,
      namespace,
      replicas: apiReplicas,
      serviceAccountName,
      nodeSelector: nodeSelector,
      initContainers: [],
      containers: [
        {
          name: apiName,
          image,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          ports: [
            {
              containerPort: 8000,
            },
          ],
          resources: apiResources,
          envFrom: applicationEnv,
          readinessProbe: {
            httpGet: {
              port: IntOrString.fromNumber(8000),
              path: "/v1/status",
            },
            initialDelaySeconds: 10,
          },
          livenessProbe: {
            httpGet: {
              port: IntOrString.fromNumber(8000),
              path: "/v1/status",
            },
            initialDelaySeconds: 10,
          },
        },
      ],
      servicePort: 80,
      containerPort: 8000,
    });

    if (segment === Segment.LOC00) {
      /**
       * Create ingress
       */
      new KubeIngress(this, `${name}-ingress`, {
        metadata: {
          name: `${name}-ingress`,
          namespace,
          annotations: {
            "nginx.ingress.kubernetes.io/enable-cors": "true",
            "nginx.ingress.kubernetes.io/cors-allow-origin":
              "https://apps.bettermarks.loc",
          },
        },
        spec: {
          ingressClassName: LICENSING_INGRESS_CLASS,
          rules: [
            {
              http: {
                paths: [
                  {
                    path: "/",
                    pathType: HttpIngressPathType.PREFIX,
                    backend: {
                      service: {
                        name: licensingService.service!.name,
                        port: {
                          number: 80,
                        },
                      },
                    },
                  },
                ],
              },
            },
          ],
        },
      });
    }
  }
}
