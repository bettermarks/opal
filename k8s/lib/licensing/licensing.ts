import {
  EnvFromSource,
  IntOrString,
  ResourceRequirements,
} from "../../imports/k8s";
import { LicensingService } from "../common/service";
import { LICENSING_SECRET } from "../constants";
import { Namespace, NodeSelector, Segment } from "../types";
import { Chart, ChartProps } from "cdk8s";
import { ImagePullPolicy } from "cdk8s-plus-29";
import { Construct } from "constructs";

/**
 * This class is the implementation detail of Licensing deployment.
 */
export type LicensingChartProps = ChartProps & {
  /**
   * Docker image for Licensing
   */
  image: string;
  serviceAccountName: string;
  segment: Segment;
  name: string;
  /**
   * Name of the config map containing necessary configuration.
   * Keys are derived from the type ApplicationConfig
   */
  configMap: string;
  nodeSelector?: NodeSelector;
  apiResources?: ResourceRequirements;
  /**
   * API replicas
   * @default 3
   */
  apiReplicas?: number;
};

/**
 * Deploys Licensing API server
 */
export class LicensingChart extends Chart {
  constructor(scope: Construct, id: string, props: LicensingChartProps) {
    super(scope, id, props);

    const {
      namespace = Namespace.LICENSING,
      apiReplicas = 1,
      apiResources,
      image,
      nodeSelector,
      serviceAccountName,
      name,
      configMap,
    } = props;

    const applicationEnv: EnvFromSource[] = [
      {
        configMapRef: {
          name: configMap,
        },
      },
      {
        secretRef: {
          name: LICENSING_SECRET,
        },
      },
    ];

    const apiName = `${name}-api`;
    new LicensingService(this, `${apiName}-service`, {
      name: apiName,
      namespace,
      replicas: apiReplicas,
      serviceAccountName,
      nodeSelector: nodeSelector,
      initContainers: [],
      containers: [
        {
          name: apiName,
          image,
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
          ports: [
            {
              containerPort: 8000,
            },
          ],
          resources: apiResources,
          envFrom: applicationEnv,
          readinessProbe: {
            httpGet: {
              port: IntOrString.fromNumber(8000),
              path: "/status",
            },
            timeoutSeconds: 3,
            initialDelaySeconds: 10,
          },
          livenessProbe: {
            httpGet: {
              port: IntOrString.fromNumber(8000),
              path: "/livez",
            },
            timeoutSeconds: 3,
            initialDelaySeconds: 10,
          },
        },
      ],
      servicePort: 80,
      containerPort: 8000,
    });
  }
}
