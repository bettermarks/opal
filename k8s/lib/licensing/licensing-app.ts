import { APPLICATION_CONFIG, DEPLOYMENT_CONFIG } from "../common/config";
import { LocalIngressChart } from "../common/local-ingress";
import { LicensingSecrets } from "../common/secrets";
import { LicensingServiceAccount } from "../common/service-account";
import {
  APP_NODE_POOL_LABELS,
  POSTGRES_IMAGE,
  REGISTRY_CREDENTIALS,
} from "../constants";
import { PostgresChart } from "../postgres/postgres";
import { AppSynthProps, Segment } from "../types";
import { LicensingChart } from "./licensing";
import { LicensingConfig } from "./licensing-configmap";
import { EventExportCronJob } from "./licensing-event-export-cron-job";
import { MigrationJobChart } from "./licensing-migration-job";
import { App } from "cdk8s";

export const createLicensingApp = (app: App, appProps: AppSynthProps) => {
  const { segment, imageTag, namespace } = appProps;

  let pgChart: PostgresChart | undefined = undefined;
  let localIngressChart: LocalIngressChart | undefined = undefined;

  const licensingServiceAccount = new LicensingServiceAccount(
    app,
    "licensing-service-account",
    {
      imagePullSecrets: segment === Segment.LOC00 ? [] : [REGISTRY_CREDENTIALS],
      name: "licensing",
      namespace,
    },
  );
  const licensingConfig = new LicensingConfig(app, "licensing-config", {
    appConfig: APPLICATION_CONFIG[segment],
    name: "licensing-config",
    namespace,
  });
  const licensingSecrets = new LicensingSecrets(app, "licensing-secrets", {
    segment,
    namespace,
  });
  const image =
    segment === Segment.LOC00
      ? "licensing"
      : `676249682729.dkr.ecr.eu-central-1.amazonaws.com/licensing:${imageTag}`;
  const nodeSelector =
    segment === Segment.LOC00 ? undefined : APP_NODE_POOL_LABELS;
  const {
    apiResources,
    apiReplicas,
    migrationJobResources,
    eventExportResources,
  } = DEPLOYMENT_CONFIG[segment];

  const licensingChart = new LicensingChart(app, "licensing", {
    name: "licensing",
    image,
    configMap: licensingConfig.configMap.name,
    apiReplicas,
    apiResources,
    namespace,
    nodeSelector,
    segment,
    serviceAccountName: licensingServiceAccount.serviceAccount.name,
  });

  const licensingMigrationJob = new MigrationJobChart(app, "migration", {
    configMap: licensingConfig.configMap.name,
    image,
    name: "licensing",
    namespace,
    nodeSelector,
    serviceAccountName: licensingServiceAccount.serviceAccount.name,
    migrationJobResources,
  });

  const logFormat = segment === Segment.LOC00 ? "console" : "json";
  const licensingEventExportCronjob = new EventExportCronJob(
    app,
    "event-export",
    {
      configMap: licensingConfig.configMap.name,
      image,
      logFormat,
      name: "licensing-event-export",
      namespace,
      nodeSelector,
      resources: eventExportResources,
      serviceAccountName: licensingServiceAccount.serviceAccount.name,
    },
  );

  if (segment === Segment.LOC00) {
    pgChart = new PostgresChart(app, "postgres", {
      image: POSTGRES_IMAGE,
      name: "licensing-db",
      namespace,
    });
    localIngressChart = new LocalIngressChart(app, "local-ingress", {
      namespace,
    });
  }
  return {
    licensingServiceAccount,
    licensingConfig,
    licensingSecrets,
    licensingChart,
    licensingMigrationJob,
    licensingEventExportCronjob,
    pgChart,
    localIngressChart,
  };
};
