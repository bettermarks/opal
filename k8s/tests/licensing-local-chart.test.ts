import { LicensingChart } from "../lib/licensing/licensing";
import { Segment } from "../lib/types";
import { Testing } from "cdk8s";

describe("licensing-local-chart", () => {
  beforeEach(() => {
    jest.useFakeTimers().setSystemTime(new Date("2008-12-01"));
  });
  test("should match with snapshot", () => {
    const app = Testing.app();
    const chart = new LicensingChart(app, "test-chart", {
      name: "licensing",
      image: "licensing",
      namespace: "dummy-namespace",
      segment: Segment.LOC00,
      configMap: "licensing-configmap",
      serviceAccountName: "dummy-service-account",
    });
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });
});
