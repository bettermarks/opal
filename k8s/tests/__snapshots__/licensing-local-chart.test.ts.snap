// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`licensing-local-chart should match with snapshot 1`] = `
[
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "name": "licensing-api",
      "namespace": "dummy-namespace",
    },
    "spec": {
      "replicas": 1,
      "selector": {
        "matchLabels": {
          "app": "licensing-api",
        },
      },
      "strategy": {
        "rollingUpdate": {
          "maxSurge": 1,
          "maxUnavailable": 0,
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "licensing-api",
          },
        },
        "spec": {
          "containers": [
            {
              "envFrom": [
                {
                  "configMapRef": {
                    "name": "licensing-configmap",
                  },
                },
                {
                  "secretRef": {
                    "name": "licensing",
                  },
                },
              ],
              "image": "licensing",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": {
                "httpGet": {
                  "path": "/livez",
                  "port": 8000,
                },
                "initialDelaySeconds": 10,
                "timeoutSeconds": 3,
              },
              "name": "licensing-api",
              "ports": [
                {
                  "containerPort": 8000,
                },
              ],
              "readinessProbe": {
                "httpGet": {
                  "path": "/status",
                  "port": 8000,
                },
                "initialDelaySeconds": 10,
                "timeoutSeconds": 3,
              },
            },
          ],
          "initContainers": [],
          "serviceAccountName": "dummy-service-account",
        },
      },
    },
  },
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "name": "licensing-api",
      "namespace": "dummy-namespace",
    },
    "spec": {
      "ports": [
        {
          "port": 80,
          "targetPort": 8000,
        },
      ],
      "selector": {
        "app": "licensing-api",
      },
      "type": "ClusterIP",
    },
  },
]
`;
